python/mspasspy/db/database.py

```diff
@@ -291,57 +358,62 @@ def _save_elog(self, data, collection='error_logs'):
         elog is the error log object to be saved.
         Return:  List of ObjectID of inserted
         """
-        if 'elog_ids' not in data:
-            data['elog_ids'] = []
+        if not collection:
+            collection = self.database_schema.default_name('elog')
+
+        if not elog_id:
+            elog_id = []
+
         oid = None
-        if '_id' in data:
-            oid = data['_id']
-        elog = data.elog
+        if '_id' in mspass_object:
+            oid = mspass_object['_id']
+
+        if isinstance(mspass_object, Seismogram):
+            wf_id_name = 'wf_Seismogram_id'
+        elif isinstance(mspass_object, TimeSeries):
+            wf_id_name = 'wf_TimeSeries_id'
+        else:
+            raise TypeError("only TimeSeries and Seismogram are supported")
+
+        elog = mspass_object.elog
         n = elog.size()
-        if n == 0:
-            return
-        errs = elog.get_error_log()
-        jobid = elog.get_job_id()
-        oidlst = []
-        for i in range(n):
-            x = errs[i]
-            docentry = {'job_id': jobid, 'algorithm': x.algorithm, 'badness': str(x.badness),
-                        'error_message': x.message, 'process_id': x.p_id}
-            if oid:
-                docentry['wf_id'] = oid
-            oidlst.append(self[collection].insert_one(docentry).inserted_id)
-        data['elog_ids'].extend(oidlst)
+        if n != 0:
+            errs = elog.get_error_log()
+            jobid = elog.get_job_id()
+            for i in range(n):
+                x = errs[i]
+                docentry = {'job_id': jobid, 'algorithm': x.algorithm, 'badness': str(x.badness),
+                            'error_message': x.message, 'process_id': x.p_id}
+                if oid:
+                    docentry[wf_id_name] = oid
+                elog_id.append(self[collection].insert_one(docentry).inserted_id)
+        return elog_id
 
     @staticmethod
-    def _read_data_from_dfile(d):
-        if ('dir' not in d) or ('dfile' not in d) or ('foff' not in d):
-            raise KeyError("one of the keys: dir, dfile, foff is missing")
-        di = d.get_string('dir')
-        dfile = d.get_string('dfile')
-        foff = d.get('foff')
-        fname = os.path.join(di, dfile)
+    def _read_data_from_dfile(data, dir, dfile, foff):
+        fname = os.path.join(dir, dfile)
         with open(fname, mode='rb') as fh:
```

Comment:
Note that the open command can also throw exceptions, and we need to properly address that in the outer methods.